cmds/vbe.c

```diff
@@ -0,0 +1,349 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * VESA VBE modesetting
+ *
+ * Copyright 2024 Phoenix Systems
+ * Author: Adam Greloch
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "cmd.h"
+
+#include <hal/hal.h>
+#include <lib/lib.h>
+#include <syspage.h>
+
+
+#define PTR_TO_SEG(ptr)       ((addr_t)ptr >> 4)
+#define SEGOFS_TO_PTR(segofs) ((addr_t)((segofs & 0xffff0000) >> 12) + (segofs & 0xffff))
+
+#define SUPPORTS_LFB (0x80)
+
+
+typedef struct {
+	u16 width;
+	u16 height;
+	u16 bpp;
+} video_mode_t;
+
+
+/* Structure definitions follow "VESA BIOS EXTENSION (VBE) Core Functions
+ * Standard Version: 3.0", Video Electronics Standards Association, 1998 */
+
+typedef struct {
+	char VbeSignature[4];
+	u16 VbeVersion;
+	u32 OemStringPtr;
+	u32 Capabilities;
+	u32 VideoModePtr;
+	u16 TotalMemory;
+
+	u16 OemSoftwareRev;
+	u32 OemVendorNamePtr;
+	u32 OemProductNamePtr;
+	u32 OemProductRevPtr;
+	char Reserved[222];
+
+	char OemData[256];
+} __attribute__((packed)) vbe_info_block_t;
+
+
+typedef struct {
+	/* Mandatory information for all VBE revisions */
+	u16 ModeAttributes;
+	u8 WinAAttributes;
+	u8 WinBAttributes;
+	u16 WinGranularity;
+	u16 WinSize;
+	u16 WinASegment;
+	u16 WinBSegment;
+	u32 WinFuncPtr;
+	u16 BytesPerScanLine;
+
+	/* Mandatory information for VBE 1.2 and above */
+	u16 XResolution;
+	u16 YResolution;
+	u8 XCharSize;
+	u8 YCharSize;
+	u8 NumberOfPlanes;
+	u8 BitsPerPixel;
+	u8 NumberOfBanks;
+	u8 MemoryModel;
+	u8 BankSize;
+	u8 NumberOfImagePages;
+	u8 Reserved0;
+
+	u8 RedMaskSize;
+	u8 RedFieldPosition;
+	u8 GreenMaskSize;
+	u8 GreenFieldPosition;
+	u8 BlueMaskSize;
+	u8 BlueFieldPosition;
+	u8 RsvdMaskSize;
+	u8 RsvdFieldPosition;
+	u8 DirectColorModeInfo;
+
+	/* Mandatory information for VBE 2.0 and above */
+	u32 PhysBasePtr;
+	u8 Reserved1[212];
+} __attribute__((packed)) vbe_mode_info_block_t;
+
+
+static struct {
+	video_mode_t preferred;
+	video_mode_t fallback;
+	vbe_info_block_t *info;
+	vbe_mode_info_block_t *modeInfo;
+} vbe_common;
+
+
+static void _vbe_commonInit(void)
+{
+	vbe_common.preferred.height = 1280;
+	vbe_common.preferred.width = 800;
+	vbe_common.preferred.bpp = 32;
+
+	vbe_common.fallback.height = 1024;
+	vbe_common.fallback.width = 768;
+	vbe_common.fallback.bpp = 32;
+
+	/* ADDR_VBE_INFO, ADDR_VBE_MODE_INFO should be aligned to 0x10, so that they can
+	 * be referenced from 8086 emulation mode using segment register only (es=(ADDR >> 4), di=0) */
+	vbe_common.info = (vbe_info_block_t *)ADDR_VBE_INFO;
+	vbe_common.modeInfo = (vbe_mode_info_block_t *)ADDR_VBE_MODE_INFO;
+}
```

Comment:
There's merit for this initialization - otherwise we will reset these values on each cmd invocation, I don't think we want that 
```suggestion
} vbe_common = {
.preffered = { .height ....
```